### ISSUE TITLE
[RFC] - Geo Hash Grid Aggregation on Geo Shape Field

### ISSUE THREAD

**Issue by navneet1v** [opensearch-project/geospatial#193]
The purpose of this RFC (request for comments) is to gather community feedback on a proposal to allow OpenSearch users to facilitate Geo Hash Grid Aggregations over the GeoShape data type. This RFC is in continuation with #84.

## Geo Hash Grid Aggregation

GeoHash Grid is a multi-bucket aggregation which will grouping the GeoShapes into different buckets where each bucket represent the cells in a grid. Each cell is labeled using a [geohash](https://en.wikipedia.org/wiki/Geohash) which is of user-definable precision. The working of the aggregation is very similar to GeoHash Grid geo points Aggregation but there is key difference. A geo_point(if it is not multi geo_point) is only present in one bucket, but the a geo_shape will be counted in all the geo hashes grid cells with which the shape is intersecting.

### Aggregation Schema

```
{
  "aggs/aggregations": {
    "<user-provided-aggregation-name>": {
      "geohash_grid": {
        "field": "<field-name-on-which-aggregation-will-be-performed>",
        "precision" : 10, // default value is 5
        "bounds": {  // optional object
          "top_left": <Point representing the top left corner>,
          "bottom_right": <Point representing the bottom right corner>
        }
      }
    }
  }
}

```
**Input Parmeters:**

* **precision:** It is defined as the string length of the GeoHash bucket in the output. Default value is 5.
* **bounds:** It is an optional object, which bounds the region over which the aggregation is happening. Shapes which are intersecting with this bound or completely present in this bounds will only be included in the aggregation output.

**Example**

**Index Creation**

```
PUT /example
{
    "mappings": {
        "properties": {
            "location": {
                "type": "geo_shape"
            }
        }
        
    }
}
```

**Index Data**

```
POST /example/_bulk?refresh
{"index":{"_id":191}}
{"name": "NEMO Science Museum","location": {"type": "envelope","coordinates": [ [100.0, 1.0], [101.0, 0.0] ]}}
{"index":{"_id":219}}
{"name": "NEMO Science Museum","location": {"type": "envelope","coordinates": [ [100.0, 1.0], [106.0, 0.0] ]}}
```

**Run Aggregation**
```
POST /myexample/_search?size=0
{
  "aggregations": {
    "myaggregation": {
      "geohash_grid": {
        "field": "location",
        "precision": 3
      }
    }
  }
}
```

**Output of Aggregation**
```
{
....
  "aggregations": {
        "myaggregation": {
            "buckets": [
                {
                    "key": "w0p",
                    "doc_count": 2
                },
                {
                    "key": "w25",
                    "doc_count": 1
                },
                {
                    "key": "w24",
                    "doc_count": 1
                },
                {
                    "key": "w21",
                    "doc_count": 1
                },
                {
                    "key": "w20",
                    "doc_count": 1
                }
            ]
        }
    }
```

## References
1. RFC for Aggregations: #84 
2. https://github.com/opensearch-project/geospatial/issues/191

**Comment by navneet1v** [comment]
The feature is pushed in the 2.x branch of OpenSearch core. Closing this RFC

**Comment by ohltyler** [comment]
@navneet1v is this part of 2.9? If not, can it be re-labeled with the target version?

**Comment by heemin32** [comment]
It is part of 2.9. Closing it.

**Comment by navneet1v** [comment]
@ohltyler I will this issue once 2.9 is released. There is frontend work but we will track it separately.

**Comment by navneet1v** [comment]
Closing this RFC as the code is merged to 2.9 branch
